### PR TITLE
fork rust dlc crates

### PR DESCRIPTION
**Pull request by bennyhodl**
- **dep: add dlc, dlc-messages, & dlc-trie to dlcdevkit**
- **feat: splice DLC to ddk-manager**

This  PR pulls in `rust-dlc/{dlc, dlc-messages, dlc-trie}` to dlcdevkit

The decision was made after slow release times and some related features to improve `dlcdevkit`

Along with pulling in the crates it includes new features that are ported from my fork of rust-dlc

[single funded](https://github.com/bennyhodl/rust-dlc/pull/3)
[splicing](https://github.com/bennyhodl/rust-dlc/pull/4)

### CODE REVIEW

File: ddk-manager/src/dlc_input.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    let contract = storage
+        .get_contract(&contract_id)
+        .await?
+        .ok_or(Error::StorageError(
+            "Contract not found to sign DLC input.".to_string(),
+        ))?;
+
+    let key_id = match contract {
+        Contract::Confirmed(c) => Ok(c.accepted_contract.offered_contract.keys_id),
+        _ => Err(Error::InvalidState(
```
Comment (matthewjablack) [inline ddk-manager/src/dlc_input.rs:49]: Might be good to change this to Signed (or have a state PreConfirmed) because if you initially created a single funding transaction with no change, then you have no way to fee bump, other than splice in more collateral as a CPFP

**Review by matthewjablack — COMMENTED** [review]
Do we need a `calculateMaxCollateral` from utxos?

https://github.com/AtomicFinance/bitcoin-abstraction-layer/pull/172/files#diff-35671a6241dccb1b562d72bb93c9632079655b502dd85c567f0bf90e489d2e85R572-R589

File: ddk-manager/src/dlc_input.rs (hunk 1 above)
Comment (matthewjablack) [reply ddk-manager/src/dlc_input.rs:49]: On second thought, this seems unnecessary unless it becomes a large problem down the road

File: ddk-manager/src/manager.rs [hunk 2]
```diff
@@ -487,11 +544,50 @@ where
...
+                signed_cet: signed_contract
+                    .accepted_contract
+                    .dlc_transactions
+                    .fund
+                    .clone(),
+            };
+
+            self.store
+                .update_contract(&Contract::PreClosed(preclosed_contract.clone()))
+                .await?;
```
Comment (matthewjablack) [inline ddk-manager/src/manager.rs:588]: What happens if splice transaction ends up being RBF'd? 

I'm assuming this doesn't block it from changing to Close somewhere else with an execution transaction

File: dlc/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,1722 @@
...
+
+/// Contains the necessary transactions for establishing a DLC
+#[derive(Clone)]
+#[cfg_attr(feature = "use-serde", derive(Serialize, Deserialize))]
+pub struct DlcTransactions {
+    /// The fund transaction locking both parties collaterals
+    pub fund: Transaction,
+    /// The contract execution transactions for closing the contract on a
+    /// certain outcome
+    pub cets: Vec<Transaction>,
```
Comment (matthewjablack) [inline dlc/src/lib.rs:131]: To confirm, a splice tx is just added to cets correct (in the DlcTransactions object)?

File: ddk-manager/tests/splice_execution_tests.rs [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+        }
+    };
+
+    let alice_splice_offer = alice_manager_send
+        .lock()
+        .await
+        .send_splice_offer(
+            &test_params.contract_input,
+            spliced_input.bob_public_key,
+            &spliced_input.contract_id,
```
Comment (matthewjablack) [inline ddk-manager/tests/splice_execution_tests.rs:353]: Edge case, what if we try to splice the same contract twice?

Comment (matthewjablack) [reply ddk-manager/tests/splice_execution_tests.rs:353]: Also would be good to have a test that references a contract ID that doesn't exist

i.e. fake_contract_id

Comment (matthewjablack) [reply ddk-manager/tests/splice_execution_tests.rs:353]: And maybe a test with collateral below dust as well (to make sure that fails)

**Review by matthewjablack — COMMENTED** [review]
Looking solid, just a couple edge cases for tests that could be added

**Review by matthewjablack — APPROVED** [review]
Tests LGTM 🎉
